Improve fmpq interpolation

rprebet opened the pull request:
### What's in this PR
This PR introduces two new implementations of Lagrange interpolation for rational polynomials (`fmpq_poly`) that avoid denominator growth and rely purely on integer arithmetic.

The two routines are:
- a naive version based on direct formula evaluation;
- a fast version using a subproduct tree structure for efficient polynomial operation.

Both avoid introducing denominators during intermediate computations, keeping integer arithmetic throughout.

Following the TODO, I added for the naive version, additional function accepting inputs `(fmpq *xs, fmpq *ys)` and `(fmpz *xs, fmpq *ys)`. However, these variants internally convert to rational format and call the main general routine
Indeed, tailored versions don't seem to bring meaningful performance improvements.

According to the timings below, the fast version is almost consistently faster than the naive version. I included the naive version in case, to be consistent with the rest of FLINT, and since it seems slightly better for small degree and coefficient size.

### What's not in this PR
The choice of these functions has been made after several performance experiments. In particular:
- Newton interpolation (as suggested in the TODO) was implemented but proved significantly slower due to increasing denominator sizes. I omit it in favor of the generic ring fallback.
- A multi-modular version of the interpolation was also implemented (using Chinese Remaindering and rational reconstruction). However, it remains >2-3x slower than the naive above implementation. The slowdown is due to the heavier bit sizes in rational interpolation which requires to several prime numbers. That version is not part of this PR but could be explored for large-scale inputs with some optimizations (I did not have to time to add these yet).

Note that I also implemented rational multipoint evaluation (with the same tricks) , but it turned out much slower (even in the integer-only case) than iterative evaluation. This optimization was not included in the PR, but is available if needed.

I can also add a `fmpz_poly_pseudodiv_root_fmpq` if requested, when the division is not exact.

### Questions
About the `(fmpq *xs, fmpq *ys)` and `(fmpz *xs, fmpq *ys)` versions:
- Should we keep these versions?
- If yes, should we have similar ones for the fast method?
- Should we also test these versions ?

Is the renaming of fmpz_poly_div_root → fmpz_poly_div_root_fmpz ok?
I'm not sure about the copyright/license convention here. I tried to mimic what was done elsewhere.

Many thanks for your comments!

```
Timings: 
n: nb of interpolation points  
ht(x): max height of these points


n        ht(x)     naive          fast           speedup
10       10        0.000232674    4.55258e-05     5.11x
20       10        0.000118888    0.000130761     0.91x
40       10        0.000567044    0.00102483      0.55x
80       10        0.00626858     0.00369953      1.69x
160      10        0.0554539      0.0290124       1.91x
320      10        0.892892       0.179523        4.97x
10       100       0.000134839    0.000117875     1.14x
20       100       0.00131301     0.00119543      1.10x
40       100       0.0259431      0.0163317       1.59x
80       100       0.502813       0.200439        2.51x
160      100       13.1607        2.38795         5.51x
320      100       293.021        28.6824         10.22x
10       1000      0.00375532     0.00347766      1.08x
20       1000      0.0744081      0.0519205       1.43x
40       1000      1.2538         0.516031        2.43x
80       1000      20.654         4.20887         4.91x
160      1000      422.555        42.7671         9.88x
```

albinahlback commented:
Do you have the profiling code?

fredrik-johansson commented:
This looks good, but I think you should profile y values that are random polynomials evaluated at the x values, in addition to random y values. With random y values the interpolant will always have big coefficients, but CRT (with early termination) should do well when the interpolant has small coefficients.

fredrik-johansson commented:
> Note that I also implemented rational multipoint evaluation (with the same tricks) , but it turned out much slower (even in the integer-only case) than iterative evaluation.

This is not surprising. Iterative evaluation is already asymptotically optimal (quasilinear in the size of the output).

rprebet commented:
I added multi-modular interpolation (that is adapted from the integer case), in addition to some helper functions.

Note a a potentially critical and open to criticism modification:  to improve interpolation check for early termination I removed the final fraction reduction in the internal evaluation functions. Instead, I removed the final fraction reduction from the internal evaluation functions. Instead, I incorporated this reduction into their overlay.
This allows for a 10-100x speedup for this step, which can be a bottleneck.

Here is a new profiling with evaluation of a random polynomial at small bitsize points:
``` 
Timings: 
n: nb of interpolation points  
ht(f): max bit height of the coefficients of f
ht(y): max bit height of the values of f
=================================================================================================================
 n       ht(y)   ht(f)  naive           fast            naive/fast      mmod            naive/mmod      fast/mmod
  10        87      10   8.01277e-05     5.00703e-05       1.6x          0.000130757     0.613x          0.383x
  20       220      10   0.000139201     0.000204689      0.68x          5.18148e-05      2.69x           3.95x
  40       462      10   0.00121912      0.0019058        0.64x          0.000155888      7.82x           12.2x
  80       860      10   0.0072484       0.00910373      0.796x          0.000538008      13.5x           16.9x
 160      1689      10   0.147728        0.0669557        2.21x          0.00238602       61.9x           28.1x
 320      3447      10   3.59075         0.661462         5.43x          0.0112459         319x           58.8x
 640      6670      10   60.7768         4.51557          13.5x          0.0588782        1030x           76.7x
  10       584     100   3.97503e-05     4.65219e-05     0.854x          0.000108686     0.366x          0.428x
  20       711     100   9.95735e-05     0.000109726     0.907x          9.65065e-05      1.03x           1.14x
  40      1757     100   0.000838765     0.00128193      0.654x          0.000393908      2.13x           3.25x
  80      3104     100   0.00842458      0.00840311          1x          0.00150532        5.6x           5.58x
 160      6408     100   0.216323        0.0952148        2.27x          0.00740014       29.2x           12.9x
 320     12489     100   3.49402         0.693341         5.04x          0.0442586        78.9x           15.7x
 640     24654     100   55.1624         4.75383          11.6x          0.294857          187x           16.1x
  10      3580    1000   6.81594e-05     0.000699025    0.0975x          0.000809002    0.0843x          0.864x
  20      9523    1000   0.000441622     0.000615345     0.718x          0.00137513      0.321x          0.447x
  40     17045    1000   0.00426212      0.00576843      0.739x          0.00382898       1.11x           1.51x
  80     36269    1000   0.0301422       0.0294326        1.02x          0.0152616        1.98x           1.93x
 160     67410    1000   0.477789        0.24485          1.95x          0.0840282        5.69x           2.91x
 320    131631    1000   6.38973         1.41525          4.51x          0.516221         12.4x           2.74x
 640    250653    1000   86.7628         9.69262          8.95x          4.14761          20.9x           2.34x
  10     59251   10000   0.00146406      0.00234781      0.624x          0.0197892       0.074x          0.119x
  20     96778   10000   0.00571021      0.00702424      0.813x          0.0457985       0.125x          0.153x
  40    149912   10000   0.0324311       0.0435421       0.745x          0.127634        0.254x          0.341x
  80    337636   10000   0.28902         0.301921        0.957x          0.55            0.525x          0.549x
 160    672808   10000   2.88066         1.71503          1.68x          2.38074          1.21x           0.72x
 320   1260656   10000   37.3004         10.0979          3.69x          12.1018          3.08x          0.834x
 640   2647600   10000   399.791         47.2735          8.46x          69.9813          5.71x          0.676x
  10    355002  100000   0.0186502       0.0186496           1x          0.297215       0.0627x         0.0627x
  20    908583  100000   0.119383        0.125694         0.95x          1.04942         0.114x           0.12x
  40   1460978  100000   0.550605        0.66525         0.828x          3.7034          0.149x           0.18x
  80   3316017  100000   4.11154         4.23018         0.972x          16.645          0.247x          0.254x
 160   6746476  100000   33.3099         21.3796          1.56x          64.3444         0.518x          0.332x
 320  12240968  100000   316.978         89.2912          3.55x          247.884          1.28x           0.36x
=================================================================================================================
```

and the corresponding profiling code:
``` C
#include <flint/profiler.h>
#include "flint/fmpq.h"
#include <flint/fmpz_vec.h>
#include <flint/fmpq_vec.h>
#include <flint/fmpq_poly.h>
#include "longlong.h"


ulong _fmpq_vec_max_height_ui(fmpq * vec, slong n) {
    ulong h = 0;

    for (slong i = 0; i < n; i++)
        h = FLINT_MAX(h, fmpq_height_bits(vec + i));

    return h;
}

void generate_unique_fmpq_vec(fmpq *xs, fmpq *ys, slong n, slong nbits, flint_rand_t state)
{
    slong i = 0;
    fmpq_t tmp;
    fmpq_init(tmp);

    while (i < n)
    {
        fmpq_randtest(tmp, state, nbits);
        int duplicate = 0;
        for (slong j = 0; j < i; j++)
        {
            if (fmpq_equal(tmp, xs + j))
            {
                duplicate = 1;
                break;
            }
        }
        if (!duplicate)
        {
            fmpq_set(xs + i, tmp);
            fmpq_randtest(ys + i, state, nbits);
            i++;
        }
    }

    fmpq_clear(tmp);
}

void generate_random_polynomial(fmpq_poly_t poly, slong n, slong nbits, flint_rand_t state)
{
    fmpq_t t;
    fmpq_init(t);
    fmpq_poly_zero(poly);
    fmpq_poly_fit_length(poly, n);
    for (slong i = 0; i < n; i++) {
        fmpq_randtest(t, state, nbits);
        fmpq_poly_set_coeff_fmpq(poly, i, t);
    }
    _fmpq_poly_set_length(poly, n);
    _fmpq_poly_normalise(poly);
}

int main(int argc, char* argv[])
{
    flint_rand_t state;
    flint_rand_init(state);

flint_printf("  n       ht(y)   ht(f)  naive \t\t fast \t\t naive/fast \t mmod \t\t naive/mmod \t fast/mmod\n");
    for (slong nbits = 10; nbits <= 100000; nbits *= 10) {
        for (slong n = 10; n <= 1000; n *= 2) {

            fmpq_poly_t poly1, poly2, poly3, f;
            fmpq_poly_init(poly1);
            fmpq_poly_init(poly2);
            fmpq_poly_init(poly3);
            fmpq_poly_init(f);

            fmpq *xs = _fmpq_vec_init(n);
            fmpq *ys = _fmpq_vec_init(n);

            generate_unique_fmpq_vec(xs, ys, n, 10, state);
            generate_random_polynomial(f, n, nbits, state);
            for (slong i = 0; i < n; i++)
                fmpq_poly_evaluate_fmpq(ys + i, f, xs + i);

            init_all_clocks();

            // Classic
            start_clock(0);
            fmpq_poly_interpolate_fmpq_vec(poly1, xs, ys, n);
            stop_clock(0);

            // Fast
            start_clock(1);
            fmpq_poly_interpolate_fast(poly2, xs, ys, n);
            stop_clock(1);

            // Multi-modular
            start_clock(2);
            fmpq_poly_interpolate_multi_mod(poly3, xs, ys, n);
            stop_clock(2);

            int match = fmpq_poly_equal(poly1, poly2) || fmpq_poly_equal(poly1, poly3);
            if (!match)
                flint_printf("No match for (%ld, %ld)\n", n, nbits);

            flint_printf("%4ld  %8ld  %6ld \t %4g \t %4g \t %5.3gx \t %4g \t %5.3gx \t %5.3gx\n",
            n, _fmpq_vec_max_height_ui(ys, n), nbits,
            get_clock(0) / 1000000, get_clock(1) / 1000000, get_clock(0) / get_clock(1),
            get_clock(2) / 1000000, get_clock(0) / get_clock(2), get_clock(1) / get_clock(2));

            _fmpq_vec_clear(xs, n);
            _fmpq_vec_clear(ys, n);
            fmpq_poly_clear(poly1);
            fmpq_poly_clear(poly2);
            fmpq_poly_clear(poly3);
            fmpq_poly_clear(f);

        }
    }
    flint_rand_clear(state);
    return 0;
}
```

```

fredrik-johansson commented:
Two comments about the multi_mod version:

* You can potentially speed up the rational reconstruction by using the denominators that have already been computed to get smaller denominators to reconstruct for consecutive coefficients. See ``fmpq_mat_set_fmpz_mat_mod_fmpz``.

* Is ``bound_bits`` really correct? In any case, interpolation over Q should be guaranteed to succeed if ``_fmpq_vec_has_unique_entries`` succeeds, so ``bound_bits`` is actually useless (if ``M`` ever exceeds ``2^(bound_bits+1)``, there is a bug in the algorithm).

rprebet commented:
Thanks for your feedback! 

>     * You can potentially speed up the rational reconstruction by using the denominators that have already been computed to get smaller denominators to reconstruct for consecutive coefficients. See `fmpq_mat_set_fmpz_mat_mod_fmpz`.

I actually did implement this at first, but then the second rational reconstruction always fail and the interpolation does not terminate. It is either a problem in my implementation (see the code below) or because the coefficients do not share a "small common denominator" (Assumption 1 in https://www.sciencedirect.com/science/article/pii/S0747717116300773).

``` C
/* Rational reconstruction */
fmpz_one(t);
rat_rec = 1;
for (i = 0; rat_rec && i < n; i++)
{
    fmpz_mul(rescoeffs + i, rescoeffs + i, t);
    fmpz_mod(rescoeffs + i, rescoeffs + i, M);
    rat_rec = fmpq_reconstruct_fmpz(coeffs + i, rescoeffs + i, M);
    fmpz_mul(fmpq_denref(coeffs + i), fmpq_denref(coeffs + i), t);
    fmpz_mul(t, fmpq_denref(coeffs + i), t);
}
```

>     * Is `bound_bits` really correct? In any case, interpolation over Q should be guaranteed to succeed if `_fmpq_vec_has_unique_entries` succeeds, so `bound_bits` is actually useless (if `M` ever exceeds `2^(bound_bits+1)`, there is a bug in the algorithm).

Indeed, it seems that I might have underestimated the bound. I agree that if the algorithm is correct, we should not exceed this bound, but how do we prevent infinite loop in case a bug have escaped our vigilance ?

> BTW, I noticed a performance bug in the existing _fmpq_poly_evaluate_fmpz when looking at your polynomial evaluation changes: this currently calls _fmpz_poly_evaluate_horner_fmpz but should definitely call _fmpz_poly_evaluate_fmpz!

i just corrected it.

fredrik-johansson commented:
> Thanks for your feedback!
> 
> > * You can potentially speed up the rational reconstruction by using the denominators that have already been computed to get smaller denominators to reconstruct for consecutive coefficients. See `fmpq_mat_set_fmpz_mat_mod_fmpz`.
> 
> I actually did implement this at first, but then the second rational reconstruction always fail and the interpolation does not terminate. It is either a problem in my implementation (see the code below) or because the coefficients do not share a "small common denominator" (Assumption 1 in https://www.sciencedirect.com/science/article/pii/S0747717116300773). But I don't get why it works for matrices...
> 
> ```c
> /* Rational reconstruction */
> fmpz_one(t);
> rat_rec = 1;
> for (i = 0; rat_rec && i < n; i++)
> {
>     fmpz_mul(rescoeffs + i, rescoeffs + i, t);
>     fmpz_mod(rescoeffs + i, rescoeffs + i, M);
>     rat_rec = fmpq_reconstruct_fmpz(coeffs + i, rescoeffs + i, M);
>     fmpz_mul(fmpq_denref(coeffs + i), fmpq_denref(coeffs + i), t);
>     fmpz_set(t, fmpq_denref(coeffs + i));
> }
> ```

You can't overwrite ``rescoeffs`` since these need to be stable for the CRT, so it should be something like this:

```
  /* Rational reconstruction */
  fmpz_one(t);
  rat_rec = 1;
  for (i = 0; rat_rec && i < n; i++)
  {
      fmpz_mul(tmp, rescoeffs + i, t);
      fmpz_mod(tmp, tmp, M);
      rat_rec = fmpq_reconstruct_fmpz(coeffs + i, tmp, M);
      fmpq_div_fmpz(coeffs + i, coeffs + i, t);
      fmpz_set(t, fmpq_denref(coeffs + i));
  }
```

...however, I just tried this with your profile program, and it seems to make things slower, so it looks like it's not an optimization here (or one needs to apply this trick more selectively to make it an optimization).

> > * Is `bound_bits` really correct? In any case, interpolation over Q should be guaranteed to succeed if `_fmpq_vec_has_unique_entries` succeeds, so `bound_bits` is actually useless (if `M` ever exceeds `2^(bound_bits+1)`, there is a bug in the algorithm).
> 
> Indeed, it seems that I might have underestimated the bound. I agree that if the algorithm is correct, we should not exceed this bound, but how do we prevent infinite loop in case a bug have escaped our vigilance ?

If there is really a bug, an infinite loop is hardly worse than returning the wrong result (i.e. incorrectly claiming that no interpolant exist).

rprebet commented:
Oops thanks for the correction! 
It seems that the rational reconstruction cost is sufficiently negligible compared to other steps (e.g. evaluation check) in the situation where f has small coefficients. Which could explain the optimization failure.

Indeed, using the first profile program (random y values) to compare both versions, I get a speedup with this optimization:
```
n        ht(x)   old             new            speedup
10       10      0.00139574      0.00020211      6.91x
20       10      0.00363093      0.00316226      1.15x
40       10      0.0333947       0.030337        1.10x
80       10      0.16393         0.122463        1.34x
160      10      1.50931         1.09255         1.38x
320      10      8.69029         5.34263         1.63x
10       100     0.0102513       0.0108162       0.95x
20       100     0.101966        0.0745992       1.37x
40       100     1.41299         1.07323         1.32x
80       100     14.657          7.97921         1.84x
```
But in these situations, the multi-modular approach is slower than the classic and fast versions.

Besides, I removed the `bound_bits`.

fredrik-johansson commented:
BTW, when generating rational coefficients with random numerators and denominators, the common denominator will be ~= product of all denominators. Generating a random vector of integers and dividing by a random common denominator may be more representative of real-world usage.

rprebet commented:
> BTW, when generating rational coefficients with random numerators and denominators, the common denominator will be ~= product of all denominators. Generating a random vector of integers and dividing by a random common denominator may be more representative of real-world usage.

Ok, so I used the following generating function for the random polynomial:
``` C
void generate_random_polynomial_den(fmpq_poly_t poly, slong n, slong nbits, flint_rand_t state)
{
    fmpq_t t;
    fmpq_init(t);
    fmpq_poly_zero(poly);
    fmpq_poly_fit_length(poly, n);
    for (slong i = 0; i < n; i++)
        fmpz_randtest(poly->coeffs + i, state, nbits);
    do {
        fmpz_randtest_not_zero(poly->den, state, nbits);
    } while (fmpz_is_zero(poly->den)); // Just in case
    _fmpq_poly_set_length(poly, n);
    _fmpq_poly_normalise(poly);
}
```
and I get indeed different speedups:
 * multi-modular functions seem to have a better speedup w.r.t. fast function;
 * the common denominator rational reconstruction seems now to perform almost always slightly better.

This might then suggest it is an optimization in this case.  But it is not clear to me how to efficiently detect such cases.
I tried by checking if the denominator just after the reconstruction of the second coefficient (i.e. before multiplying denominator of the first coefficient) is small, but this was not very conclusive...

In the timings below, I omitted the naive approach as it did not seem relevant to me here.
```
mmod: **without** common denominator reconstruction
mmod1: **with** common denominator reconstruction
=================================================================================================================
   n       ht(y)   ht(f)    fast            mmod        mmod/fast           mmod1       fast/mmod1      mmod/mmod1
  10        74      10   6.13103e-05     9.31613e-05     0.658x          1.14781e-05      5.34x           8.12x
  20       197      10   9.55323e-05     2.68639e-05      3.56x          1.92245e-05      4.97x            1.4x
  40       389      10   0.000777732     0.000101692      7.65x          7.04142e-05        11x           1.44x
  80       778      10   0.0083626       0.000376363      22.2x          0.000285552      29.3x           1.32x
 160      1590      10   0.0607823       0.00170439       35.7x          0.00154195       39.4x           1.11x
 320      3193      10   0.523478        0.00865207       60.5x          0.00858154         61x           1.01x
 640      6397      10   5.05206         0.0541963        93.2x          0.0524969        96.2x           1.03x
1280     12794      10   31.2328         0.279288          112x          0.281939          111x          0.991x
  10       141     100   2.97039e-05     7.81968e-05      0.38x          2.9951e-05      0.992x           2.61x
  20       266     100   8.73303e-05     0.00010721      0.815x          9.24781e-05     0.944x           1.16x
  40       475     100   0.00102182      0.000253617      4.03x          0.000227321       4.5x           1.12x
  80       880     100   0.00855508      0.000850608      10.1x          0.000720193      11.9x           1.18x
 160      1687     100   0.0626588       0.00301647       20.8x          0.00270071       23.2x           1.12x
 320      3259     100   0.647697        0.0128175        50.5x          0.0117861          55x           1.09x
 640      6477     100   3.24901         0.0625034          52x          0.0612191        53.1x           1.02x
  10      1044    1000   5.4669e-05      0.000702463    0.0778x          0.000615735    0.0888x           1.14x
  20      1153    1000   0.000137572     0.00101888      0.135x          0.000997084     0.138x           1.02x
  40      1299    1000   0.00100009      0.00196101       0.51x          0.00179586      0.557x           1.09x
  80      1719    1000   0.00910127      0.00528765       1.72x          0.00397196       2.29x           1.33x
 160      2566    1000   0.0849256       0.0121967        6.96x          0.0102762        8.26x           1.19x
 320      4173    1000   0.470491        0.0289178        16.3x          0.0281185        16.7x           1.03x
 640      7374    1000   3.89268         0.112114         34.7x          0.110012         35.4x           1.02x
  10      9682   10000   0.000100362     0.0101636     0.00987x          0.0100431     0.00999x           1.01x
  20      9481   10000   0.000505616     0.0133289      0.0379x          0.0138523      0.0365x          0.962x
  40     10090   10000   0.00296239      0.0291608       0.102x          0.0268007       0.111x           1.09x
  80     10657   10000   0.0132806       0.0524002       0.253x          0.0508048       0.261x           1.03x
 160     11413   10000   0.0869277       0.114728        0.758x          0.114855        0.757x          0.999x
 320     13015   10000   0.720958        0.349957         2.06x          0.304865         2.36x           1.15x
 640     16310   10000   3.45468         0.754472         4.58x          0.72011           4.8x           1.05x
  10     83751  100000   0.00122582      0.15771       0.00777x          0.136015      0.00901x           1.16x
  20     98594  100000   0.00352496      0.25683        0.0137x          0.219006       0.0161x           1.17x
  40     96944  100000   0.020529        0.663197        0.031x          0.513209         0.04x           1.29x
  80    100324  100000   0.0711985       1.75897        0.0405x          1.21007        0.0588x           1.45x
 160     99877  100000   0.212819        2.74693        0.0775x          2.1836         0.0975x           1.26x
 320    102891  100000   1.12214         6.91446         0.162x          5.58441         0.201x           1.24x
 640    106274  100000   5.56845         14.1229         0.394x          11.4909         0.485x           1.23x
=================================================================================================================
```

fredrik-johansson commented:
Looks reasonable, though I guess there should be a max(...,0) in the expression for fbits rather than an abs(...).

BTW, there's no need to duplicate the special cases for n = 1, n = 2 in all the algorithms; it suffices have such special cases in the main interpolation function.

fredrik-johansson commented:
I'd say checking for non-unique evaluation points is desirable, but would be sufficient to implement in the main function.
